### PR TITLE
sped up scheduler tests by using fake clock

### LINKED ISSUE #98495: Try to speed up controller/nodelifecycle/scheduler unit tests
`k8s.io/kubernetes/pkg/controller/nodelifecycle/scheduler` tests are using real time delays in `pkg/controller/nodelifecycle/scheduler/timed_workers_test.go`. Example: https://github.com/kubernetes/kubernetes/blob/4b3d6369dd204c38959d4f508c7a3a3d3d81eb35/pkg/controller/nodelifecycle/scheduler/timed_workers_test.go#L64-L75 

Need to check whether they can be re-written to use fake clock instead.

Related to: #98486

/area test
/sig node

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
**What type of PR is this?**

/kind cleanup
/kind failing-test
/sig node

**What this PR does / why we need it**:

```
$ go test k8s.io/kubernetes/pkg/controller/nodelifecycle/scheduler 
ok      k8s.io/kubernetes/pkg/controller/nodelifecycle/scheduler        22.716s
```

**Which issue(s) this PR fixes**:
Fixes #98495

**Special notes for your reviewer**:

There are still tests to improve in this package, but 22 seconds may be good enough for now.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Review by ehashman — COMMENTED** [review]
```
# before
ehashman@red-dot:~/src/k8s$ go test k8s.io/kubernetes/pkg/controller/nodelifecycle/scheduler 
ok  	k8s.io/kubernetes/pkg/controller/nodelifecycle/scheduler	52.537s

# after
ehashman@red-dot:~/src/k8s$ go test k8s.io/kubernetes/pkg/controller/nodelifecycle/scheduler 
ok  	k8s.io/kubernetes/pkg/controller/nodelifecycle/scheduler	22.538s
```

/lgtm
/triage accepted
/priority important-longterm

**Comment by BenTheElder** [comment]
@mrunalp re: https://github.com/kubernetes/kubernetes/pull/98595#event-4310333462

But the node controller is still not approved? 🤔

**Comment by ehashman** [comment]
@BenTheElder yeah we approved for node, but need someone else for the final approval, either someone in the listed OWNERS file or I think a top-level approver can also do it?

**Comment by BenTheElder** [comment]
maybe @liggitt can review this for unit test speedup 🙃

File: staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go [hunk 1]
```diff
@@ -145,6 +153,29 @@ func (f *FakeClock) After(d time.Duration) <-chan time.Time {
...
+func (f *FakeClock) AfterFunc(d time.Duration, cb func()) Timer {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+	stopTime := f.time.Add(d)
+	ch := make(chan time.Time, 1) // Don't block!
+
+	go func() {
+		<-ch
+		cb()
+	}()
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go:166]: no goroutines in fake clocks... they mean that after calling time.Step(), our callback will be called asynchronously, which defeats the deterministic aspect of a fake clock. suggest something like the following instead:

```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go b/staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go
index 5346b66421a..0e42c9c00b9 100644
--- a/staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go
+++ b/staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go
@@ -103,6 +103,7 @@ type fakeClockWaiter struct {
 	stepInterval  time.Duration
 	skipIfBlocked bool
 	destChan      chan time.Time
+	afterFunc     func()
 }
 
 // NewFakePassiveClock returns a new FakePassiveClock.
@@ -160,16 +161,12 @@ func (f *FakeClock) AfterFunc(d time.Duration, cb func()) Timer {
 	stopTime := f.time.Add(d)
 	ch := make(chan time.Time, 1) // Don't block!
 
 	timer := &fakeTimer{
 		fakeClock: f,
 		waiter: fakeClockWaiter{
 			targetTime: stopTime,
 			destChan:   ch,
+			afterFunc:  cb,
 		},
 	}
 	f.waiters = append(f.waiters, timer.waiter)
@@ -241,6 +238,9 @@ func (f *FakeClock) setTimeLocked(t time.Time) {
 			} else {
 				w.destChan <- t
 			}
+			if w.afterFunc != nil {
+				w.afterFunc()
+			}
 
 			if w.stepInterval > 0 {
 				for !w.targetTime.After(t) {
```

File: staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go [hunk 2]
```diff
@@ -145,6 +153,29 @@ func (f *FakeClock) After(d time.Duration) <-chan time.Time {
 	return ch
 }
 
+// AfterFunc is the Fake version of time.AfterFunc(d, callback).
+func (f *FakeClock) AfterFunc(d time.Duration, cb func()) Timer {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go:157]: this also needs a unit test exercising the implementation. suggest something like:

```go
func TestFakeAfterFunc(t *testing.T) {
	tc := NewFakeClock(time.Now())
	if tc.HasWaiters() {
		t.Errorf("unexpected waiter?")
	}
	expectOneSecTimerFire := false
	oneSecTimerFire := 0
	tc.AfterFunc(time.Second, func() {
		if !expectOneSecTimerFire {
			t.Errorf("oneSecTimer func fired")
		} else {
			oneSecTimerFire++
		}
	})
	if !tc.HasWaiters() {
		t.Errorf("unexpected lack of waiter?")
	}

	expectOneOhOneSecTimerFire := false
	oneOhOneSecTimerFire := 0
	tc.AfterFunc(time.Second+time.Millisecond, func() {
		if !expectOneOhOneSecTimerFire {
			t.Errorf("oneOhOneSecTimer func fired")
		} else {
			oneOhOneSecTimerFire++
		}
	})

	expectTwoSecTimerFire := false
	twoSecTimerFire := 0
	twoSecTimer := tc.AfterFunc(2*time.Second, func() {
		if !expectTwoSecTimerFire {
			t.Errorf("twoSecTimer func fired")
		} else {
			twoSecTimerFire++
		}
	})

	tc.Step(999 * time.Millisecond)

	expectOneSecTimerFire = true
	tc.Step(time.Millisecond)
	if oneSecTimerFire != 1 {
		t.Errorf("expected oneSecTimerFire=1, got %d", oneSecTimerFire)
	}
	expectOneSecTimerFire = false

	expectOneOhOneSecTimerFire = true
	tc.Step(time.Millisecond)
	if oneOhOneSecTimerFire != 1 {
		// should not double-trigger!
		t.Errorf("expected oneOhOneSecTimerFire=1, got %d", oneOhOneSecTimerFire)
	}
	expectOneOhOneSecTimerFire = false

	// ensure a canceled timer doesn't fire
	twoSecTimer.Stop()
	tc.Step(time.Second)
}
```

Comment (SergeyKanzhelev) [reply staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go:158]: Thank you! Took it verbatim.

File: staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go (hunk 1 above)
Comment (SergeyKanzhelev) [reply staging/src/k8s.io/apimachinery/pkg/util/clock/clock.go]: makes sense. Applied suggested fix.

**Comment by SergeyKanzhelev** [comment]
commenting to make needs-rebase go away

**Comment by SergeyKanzhelev** [comment]
I'll try again. Failures doesn't look like related to my changes. If will fail again will investigate deeper

/retest

**Comment by SergeyKanzhelev** [comment]
internal modules verification pass locally and update does nothing. 

``` sh
$ ./hack/update-internal-modules.sh
=== tidying go.mod/go.sum in hack/tools
=== tidying go.mod/go.sum in staging/src/k8s.io/code-generator/examples
$ git status
On branch nodelifecycleSchedulerUnitTestsSpeedUp
Your branch is up to date with 'origin/nodelifecycleSchedulerUnitTestsSpeedUp'.

nothing to commit, working tree clean
```

rebased again just in case.

**Pull request merged**